src/fec_context.h

```diff
@@ -70,123 +84,98 @@ class DecodeContext {
+        inv_A_i = std::unique_ptr<vec::Vector<T>>(new vec::Vector<T>(gf, n));
+        S = std::unique_ptr<vec::Poly<T>>(new vec::Poly<T>(gf, k));
 
-        poly1_n = new vec::Poly<T>(this->gf, n);
-        poly2_n = new vec::Poly<T>(this->gf, n);
+        vec1_n = std::unique_ptr<vec::Vector<T>>(new vec::Vector<T>(gf, n));
+        vec2_n = std::unique_ptr<vec::Vector<T>>(new vec::Vector<T>(gf, n));
 
-        poly1_2k = new vec::Poly<T>(this->gf, len_2k);
-        poly2_2k = new vec::Poly<T>(this->gf, len_2k);
+        vec1_2k =
+            std::unique_ptr<vec::Vector<T>>(new vec::Vector<T>(gf, len_2k));
+        vec2_2k =
+            std::unique_ptr<vec::Vector<T>>(new vec::Vector<T>(gf, len_2k));
 
         // zero-out all polynomials as they are used in full-length for FFT
-        A->zero();
-        inv_A_i->zero();
-        S->zero();
+        A->zero_fill();
+        inv_A_i->zero_fill();
+        S->zero_fill();
 
-        poly1_n->zero();
-        poly2_n->zero();
+        vec1_n->zero_fill();
+        vec2_n->zero_fill();
     }
 
-    ~DecodeContext()
-    {
-        if (A)
-            delete A;
-        if (A_fft_2k)
-            delete A_fft_2k;
-        if (inv_A_i)
-            delete inv_A_i;
-        if (S)
-            delete S;
-        if (poly1_n)
-            delete poly1_n;
-        if (poly2_n)
-            delete poly2_n;
-    }
+    ~DecodeContext() {}
 
-    unsigned get_len_2k()
+    unsigned get_len_2k() const
     {
         return len_2k;
     }
 
-    vec::Poly<T>* get_A()
-    {
-        return A;
-    }
-
-    vec::Poly<T>* get_A_fft_2k()
-    {
-        return A_fft_2k;
-    }
-
-    vec::Poly<T>* get_inv_A_i()
-    {
-        return inv_A_i;
-    }
-
-    vec::Poly<T>* get_poly1_n()
-    {
-        return poly1_n;
-    }
-
-    vec::Poly<T>* get_poly2_n()
-    {
-        return poly2_n;
-    }
-
-    vec::Poly<T>* get_poly1_2k()
-    {
-        return poly1_2k;
-    }
-
-    vec::Poly<T>* get_poly2_2k()
-    {
-        return poly2_2k;
-    }
-
-    vec::Poly<T>* get_S()
-    {
-        return S;
-    }
-
-    vec::Vector<T>* get_frag_ids()
+    vec::Vector<T>* get_fragments_id() const
     {
         return fragments_ids;
     }
 
-    void set_frag_ids(vec::Vector<T>* fragments_ids)
-    {
-        this->fragments_ids = fragments_ids;
-    }
-
-    int get_vx_zero()
-    {
-        return vx_zero;
+    vec::Vector<T>* get_vector(unsigned vec_type) const
+    {
+        switch (vec_type) {
+        case vec_A_fft_2k:
+            return A_fft_2k.get();
+        case vec_inv_A_i:
+            return inv_A_i.get();
+        case vec_n_1:
+            return vec1_n.get();
+        case vec_n_2:
+            return vec2_n.get();
+        case vec_2k_1:
+            return vec1_2k.get();
+        case vec_2k_2:
+            return vec2_2k.get();
+        default:
+            return nullptr;
+        }
     }
 
-    void set_vx_zero(int value)
+    vec::Poly<T>* get_poly(unsigned vec_type) const
```

Comment:
If you have two set of "things" (vec and poly), then I think it would make sense to have two set of enum (CtxVec and CtxPoly)

Other than that, same remarks for get_vec.